Support chart wide and pod specific config of imagePullSecrets

consideRatio opened the pull request:
## Motivation
Before a k8s pod can starts its containers, _Kubelet_, a k8s component will download the images from the image registry. If this image registry requires credentials to access it, the Pod can specify `spec.imagePullSecrets`. If that is done, Kubelet will look in the referenced Kubernetes Secret resources and use the credentials within them to download the images.

This Helm chart both supports the creation of such k8s Secret with credentials, and the configuration of the hub pods and user pods `spec.imagePullSecrets` that reference the created Secret. What is missing though, is the same kind of configuration for all other pods.

This PR aims to resolve this once and for all by providing a chart wide configuration that augments all pods `spec.imagePullSecrets`, either with a manually created k8s Secret, or by a k8s Secret created by the Helm chart given configured credentials.

## Helm chart config changes

### Added

- `imagePullSecrets` - A place to specify manually created k8s Secret references for all pods.
- `imagePullSecret.create|registry|username|password` - A place to specify credentials for the creation of a k8s Secret that will be added to the list of all pods `spec.imagePullSecrets`.

### Updated

- `...image.pullPolicy` configuration is now available for all pods
- `...image.pullSecrets` configuration is now available for all pods

### Deprecated

- `hub.imagePullSecret` - Deprecated in favor of chart wide `imagePullSecret`
- `singleuser.imagePullSecret` - Deprecated in favor of chart wide `imagePullSecret`

## Implementation steps

- [x] Add config `imagePullSecret.create` etc, to create a k8s Secret
  resource with image registry credentials.
- [x] Define a helper function (`"jupyterhub.imagePullSecrets"`) that returns a
  JSON formatted list of Kubernetes Secret names or a blank string representing:
  the Secret created by `imagePullSecret`, the `imagePullSecrets`
  list, and the pod's specific config in `...image.pullSecrets`.
- [x] Enable all pods to use these credentials to pull images with the helper
  function.
- [x] Add a hard deprecation of `hub.imagePullSecret` and
  `singleuser.imagePullSecret`. A message will show asking the user to use
  `imagePullSecret` instead from now on.
- [x] Add config `imagePullSecrets` to allow users to specify manually
  created Kubernetes Secret's for all pods to make use of.
- [x] Add config `...image.pullSecrets` and `...image.pullPolicy` where it was
  missing for certain pods.
- [x] Update schema.yaml to reflect changes.
- [x] Update `jupyterhub_config.py` logic of the user pod's `imagePullPolicy`
  and `imagePullSecrets`
  - [x] Update KubeSpawner to support configuration where mutliple imagePullSecrets are specified, like in the k8s native specification. https://github.com/jupyterhub/kubespawner/pull/442
  - [x] Release KubeSpawner 0.14.0
  - [x] Bump KubeSpawner to 0.14.0 in this repo

Closes #981, Closes #1504, Closes #1467, Closes #1465

manics commented:
It's quite a big change to review without knowledge of exactly how this all works. Is it possible someone wants to pull their singleuser image from a different registry to the hub and proxy?

Do you have any idea how many users rely on a private registry? If it's a significant number maybe we should convert one of the existing CI jobs to test it?

consideRatio commented on jupyterhub/files/hub/jupyterhub_config.py:
```diff
@@ -163,8 +163,17 @@ def camelCaseify(s):
+if get_config('singleuser.image.pullSecrets'):
+    image_pull_secrets.extend(get_config('singleuser.image.pullSecrets'))
+if image_pull_secrets:
+    c.KubeSpawner.image_pull_secrets = image_pull_secrets
```
This is the logic that configures KubeSpawner in the same way that other pods are configured through a Helm chart helper function defined in `_helpers.tpl`, but KubeSpawner must be configured here with different syntax.

consideRatio commented on jupyterhub/schema.yaml:
```diff
@@ -1,55 +1,21 @@
 type: object
 properties:
-  hub:
+  global:
```
`schema.yaml` in its entirety is mostly just documentation, but its also some adjustment to avoid repeating the same stuff over and over by using YAML anchors (`$`) and references (`*`) to the previously defined anchors.

This section could be reviewed in the PR review by RTD. See https://zero-to-jupyterhub--1794.org.readthedocs.build/en/1794/reference/reference.html#global

consideRatio commented on jupyterhub/templates/NOTES.txt:
```diff
@@ -48,4 +48,16 @@ resources.
+{{- if .Values.singleuser.imagePullSecret.enabled }}
+{{ "HARD DEPRECATION: singleuser.imagePullSecret has renamed to global.imagePullSecret" | fail }}
+{{- end }}
+{{- end }}
```
This is the breaking changes that the user will be aware of if needed during Helm template rendering. This is because when this template renders, the `| fail` part will emit the message piped to it.

I've considered a sot deprecation, but the Helm templates logic to allow for that is messy, and we don't want to silently let users get stuck failing to pull images suddenly because their configuration isn't working any more, then its better to fail hard before any changes are made to k8s.

consideRatio commented on jupyterhub/templates/_helpers.tpl:
```diff
@@ -172,12 +172,12 @@ component: {{ include "jupyterhub.componentLabel" . }}
-{{ include "jupyterhub.dockersingleuserconfigjson.yaml" . | b64enc }}
+{{- define "jupyterhub.dockerconfigjson" -}}
+{{ include "jupyterhub.dockerconfigjson.yaml" . | b64enc }}
 {{- end }}
```
The chart previously allowed for the creation of separate secrets through the Helm chart, one for the hub pod, and one for the user pods, but now we only let them create a global or reference those they have manually created. This is what relates to the hard deprecation and `fail` command in the NOTES.txt file.

consideRatio commented on jupyterhub/templates/_helpers.tpl:
```diff
@@ -194,33 +194,33 @@ component: {{ include "jupyterhub.componentLabel" . }}
-      "auth": {{ (print .username ":" .password) | b64enc | quote }}
-    }
-  }
-}
```
Here is a section removed which was one of two almost identical sections, one for hub, and one for singleuser. The other section is made into the global version.

consideRatio commented on jupyterhub/templates/_helpers.tpl:
```diff
@@ -194,33 +194,33 @@ component: {{ include "jupyterhub.componentLabel" . }}
+
+{{- /* Write the results */}}
+{{- $_.res | toJson }}
+
```
To review this is really not possible, this requires testing locally pretty much. I used the tools/templates/lint-and-validate.py script with the associated values.

```console
# this was used a lot to get things right
helm template ./jupyterhub --values tools/templates/lint-and-validate-values.yaml --show-only templates/image-registry-credentials.yaml
helm template ./jupyterhub --vlues tools/templates/lint-and-validate-values.yaml --show-only templates/hub/deployment.yaml
```

To use Helm templates for somewhat more complicated stuff isn't fun, I hope as few people as possible ever need to learn about this.

consideRatio commented on jupyterhub/templates/hub/deployment.yaml:
```diff
@@ -59,15 +59,8 @@ spec:
-        {{- end }}
+      {{- with include "jupyterhub.imagePullSecrets" (dict "root" . "image" .Values.hub.image) }}
+      imagePullSecrets: {{ . }}
       {{- end }}
```
I embedded the logic to render a JSON formatted output such as `["secret-name-a", "secret-name-b"]` or a falsy blank string in a template helper function to avoid adding this otherwise fluffy logic in each pod.

consideRatio commented on jupyterhub/templates/image-puller/_daemonset-helper.yaml:
```diff
@@ -110,9 +97,6 @@ spec:
           image: {{ $container.image }}
-          {{- with $container.imagePullPolicy }}
-          imagePullPolicy: {{ . }}
-          {{- end }}
```
I figured that it doesn't make sense for the prePuller to have a imagePullPolicy configured like the singleuser has it configured, so I removed these from here. I'm not 100% on this, but it also reduced the complexity and I could not see it turning out to be especially bad.

consideRatio commented:
> It's quite a big change to review without knowledge of exactly how this all works. 

Yeah =/ Thank you for trying your best! I added some review comments which can potentially make it a bit easier to understand what various sections are about.

> Is it possible someone wants to pull their singleuser image from a different registry to the hub and proxy?

They can reference `singleuser.image.pullSecrets` to add references to k8s secrets in addition to anything declared on the `global.imagePullSecret(s)` level influencing all pods. This is because the logic will augment them in a list.

> Do you have any idea how many users rely on a private registry? If it's a significant number maybe we should convert one of the existing CI jobs to test it?

Not sure, but it's quite common. I've tested it quite well manually by rendering templates and ensuring the output makes sense for various pods. We could add this as a test quite easily without much complexity added by adding a dummy image to a fake-private registry with credentials that are public, and then we try it as an extra `initContainer` to some pod which we configure to use these credentials for.

I'm not sure how to go about getting such registry wich public credentials that we could make use of right now though, if I had that I could go ahead and add that to this PR as well right now.

yuvipanda reviewed (changes requested):
This is great work! I'm happy to dogfood this in one of the clusters I maintain that require imagePullSecrets.

A few comments!

1. Can we just use `imagePullSecrets` instead of `global.imagePullSecrets`? This is what we do for other global things (like rbac), and it would be nice to keep that consistent, instead of adding another layer.
2. Does this mean you can't have a separate secret for hub & user pods? I don't fully understand that, can you explain why?
3. Pre-puller modes might need more than just single-user secrets, since they might also be pulling in additional images that are from different image sources - something specified with `singleuser.extraContainers`, for example. I *think* this is also configurable here, but wanted to make sure.

In general, <3 the globalized imagePullSecrets along with per-pod overrides!

consideRatio commented:
> 1. Can we just use imagePullSecrets instead of global.imagePullSecrets? This is what we do for other global things (like rbac), and it would be nice to keep that consistent, instead of adding another layer.

Yeah, not sure about this, I considered both, but since I'm not sure, I can go ahead and change to not using global.

---

> 2. Does this mean you can't have a separate secret for hub & user pods? I don't fully understand that, can you explain why?

No you can, I have this part in the docs currently, but I'll make it clearer and reference that it will augment with hub/singleuser/etc.image.pullSecrets entries.

> This will not override or get overridden by pod specific configuration, but instead augment the pod specific configuration.

---

> 3. Pre-puller modes might need more than just single-user secrets, since they might also be pulling in additional images that are from different image sources - something specified with singleuser.extraContainers, for example. I think this is also configurable here, but wanted to make sure.
In general, <3 the globalized imagePullSecrets along with per-pod overrides!

You can only use the Helm chart to _create_ one k8s Secret that will be automatically added to appended to all pods imagePullSecrets lists. But, you can use the Helm chart to specify any additional entries to the k8s Secret names in individual or all pods imagePullSecrets specification. Individual pods within their respective image.pullSecrets config, and chart wide config in `global.imagePullSecrets`.

consideRatio commented:
@yuvipanda, similarly to a comment you made in the rbac.enabled -> rbac.create PR, you can perhaps help me decide on something.

We have `imagePullSecret.create`, but it also implies an automatic appending of the created secret to all pods spec.imagePullSecrets list. Should we add a separate flag (true by default), to separate the `create` functionality from _reference by all pods spec.imagePullSecrets` functionality? If so, what name should we use? Perhaps `create` and `referenceByAllPods`?

consideRatio commented:
@yuvipanda I've now addressed your comments in https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/1794#issuecomment-705548350 and feel happy about this PR in general.

If you think this looks good, perhaps we can merge @yuvipanda?

yuvipanda commented:
Thanks for thoroughly working on this, @consideRatio!
